osu.Game/Online/Chat/ChatLink.cs

```diff
@@ -0,0 +1,160 @@
+                                if (args.Length > 1 && int.TryParse(args[1], out int mapSetId))
+                                    beatmapSetOverlay.ShowBeatmapSet(mapSetId);
+
+                                break;
+                            case "spectate":
+                                GetUserRequest req;
+                                if (int.TryParse(args[1], out int userId))
+                                    req = new GetUserRequest(userId);
+                                else
+                                    return;
+
+                                req.Success += user =>
+                                {
+                                    chat.Game?.LoadSpectatorScreen();
+                                };
+                                api.Queue(req);
+
+                                break;
+                            default:
+                                throw new ArgumentException($"Unknown osu:// link at {nameof(ChatLink)} ({urlMatch.Groups["content"].Value}).");
+                        }
+
+                        break;
+                    case "osump":
+                        if (args.Length > 1 && int.TryParse(args[1], out int multiId))
+                            chat.Game?.LoadMultiplayerLobby(multiId);
+
+                        break;
+                    case "http":
+                    case "https":
+                        if (args[0] == "osu.ppy.sh" && args.Length > 2)
+                        {
+                            switch (args[1])
+                            {
+                                case "b":
+                                case "beatmaps":
+                                    beatmapSetOverlay.ShowBeatmap(getId(args[2]));
+                                    break;
+                                case "s":
+                                case "beatmapsets":
+                                case "d":
+                                    beatmapSetOverlay.ShowBeatmapSet(getId(args[2]));
+                                    break;
+                                default:
+                                    base.OnLinkClicked();
+                                    break;
+                            }
+                        }
+                        else
+                            base.OnLinkClicked();
+                        break;
+                    default:
+                        base.OnLinkClicked();
+                        break;
+                }
+            }
+            else
+                base.OnLinkClicked();
+
+            int getId(string input)
+            {
+                var index = input.IndexOf('#');
+                return int.Parse(index > 0 ? input.Remove(index) : input);
+            }
+        }
+
+        public string TooltipText
+        {
+            get
+            {
+                if (Url == Text)
+                    return null;
+
+                var urlMatch = getUrlMatch();

```

Comment:
really ugly that we have to do this here. going forward we will need to encapsulate the logic for each link type somehow, so it's not spread across multiple methods like this.